lambdaclass/mirra_backend
# [GH-583] Add daily rewards

## Motivation

We need a daily rewards feature.
Closes #583
More info at -> https://www.notion.so/lambdaclass/Daily-Login-9b9db0c820bd49f29b5c75753ad7205b

## Summary of changes

- [Add functions for daily rewards logic](https://github.com/lambdaclass/mirra_backend/pull/602/commits/0095ebed5a2dcafd5f4d8b0ba25cf2431c7b96ce)
- [Add daily rewards business rules json and config](https://github.com/lambdaclass/mirra_backend/pull/602/commits/80486f54a00933d6c39ae267bcb4db7d8c3a9fd2)
- [Add daily rewards fields to user table and schema](https://github.com/lambdaclass/mirra_backend/pull/602/commits/4ff2db518337e8b36f2354593b9a6e91d240bd71)
- [Add endpoint to claim current daily reward](https://github.com/lambdaclass/mirra_backend/pull/602/commits/5b32c430b519c3121dcb1899e97cf9490caf841d)
- [Add new endpoint to get the user daily reward status](https://github.com/lambdaclass/mirra_backend/pull/602/commits/045fa27aa689977f04f7b80bce8c801ff508e3fd)

## How to test it?

1. Insert a user `{:ok, user} = GameBackend.Users.find_or_create_google_user_by_email("a")`
2. `user_id = user.user.id`
3. GET request to: `/users/:user_id/claim_daily_reward` and claim the reward for today:
  a. `curl -X GET http://localhost:4001/users/:user_id/claim_daily_reward`
4. Then you can check your rewards status with:
  a. `curl -X GET http://localhost:4001/users/:user_id/get_daily_reward_status`

If you want to keep claiming rewards, you have to be on a streak. To do so you can edit the user's data:
```elixir
Users.update_user(Users.get_user(user_id) |> elem(1), %{
             last_daily_reward_claim_at: DateTime.utc_now() |> DateTime.add(-1, :day),
             last_daily_reward_claim: "day_to_claim"
           })
``` 

## Checklist
